[quant][graphmode][fx] Support dynamic quantization without calibration

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43984 [quant][graphmode][fx][api] Call fuse in prepare
* #43983 [quant][graphmode][fx] Support inplace option
* **#43952 [quant][graphmode][fx] Support dynamic quantization without calibration**

Summary:
Run weight observer for dynamic quantization before inserting quant/dequant node

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23452123](https://our.internmc.facebook.com/intern/diff/D23452123)

jerryzh168 committed:
[quant][graphmode][fx] Support dynamic quantization without calibration

Summary:
Run weight observer for dynamic quantization before inserting quant/dequant node

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support dynamic quantization without calibration"


Summary:
Run weight observer for dynamic quantization before inserting quant/dequant node

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23452123](https://our.internmc.facebook.com/intern/diff/D23452123)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support dynamic quantization without calibration"


Summary:
Run weight observer for dynamic quantization before inserting quant/dequant node

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23452123](https://our.internmc.facebook.com/intern/diff/D23452123)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support dynamic quantization without calibration"


Summary:
Run weight observer for dynamic quantization before inserting quant/dequant node

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23452123](https://our.internmc.facebook.com/intern/diff/D23452123)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support dynamic quantization without calibration"


Summary:
Run weight observer for dynamic quantization before inserting quant/dequant node

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23452123](https://our.internmc.facebook.com/intern/diff/D23452123)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support dynamic quantization without calibration"


Summary:
Run weight observer for dynamic quantization before inserting quant/dequant node

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23452123](https://our.internmc.facebook.com/intern/diff/D23452123)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support dynamic quantization without calibration"


Summary:
Run weight observer for dynamic quantization before inserting quant/dequant node

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23452123](https://our.internmc.facebook.com/intern/diff/D23452123)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support dynamic quantization without calibration"


Summary:
Run weight observer for dynamic quantization before inserting quant/dequant node

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23452123](https://our.internmc.facebook.com/intern/diff/D23452123)

[ghstack-poisoned]

supriyar commented on torch/quantization/fx/quantize.py:
```diff
@@ -266,10 +266,31 @@ def prepare(self, model, qconfig_dict, inplace=False):
+    # and run the subgraph to observe the weight
+    def _run_weight_observers(self, observed):
+        for node in observed.graph.nodes:
+            if node.op == 'call_function' and node.target in WEIGHT_INDEX_DICT:
```
How does this work for dynamic quantizable modules?

jerryzh168 replied:
for dynamic quantizable modules, the weight observer is run during conversion: e.g. https://github.com/pytorch/pytorch/blob/master/torch/nn/quantized/dynamic/modules/linear.py#L96

supriyar replied:
Can we add a comment here which mentions this?

jerryzh168 replied:
sure

supriyar commented on torch/quantization/fx/quantize.py:
```diff
@@ -266,10 +266,31 @@ def prepare(self, model, qconfig_dict, inplace=False):
+        for node in observed.graph.nodes:
+            if node.op == 'call_function' and node.target in WEIGHT_INDEX_DICT:
+                for i, node_arg in enumerate(node.args):
+                    if i in WEIGHT_INDEX_DICT[node.target]:
```
Should there be a separate dict for dynamic ops?

jerryzh168 replied:
not sure, isn't the weight index the same for static and dynamic?

supriyar replied:
I think we should only run weight observers for dynamic quantizable ops, the dict has both static and dynamic ops.

jerryzh168 replied:
this pass is only called in dynamic quantization. WEIGHT_INDEX_DICT is containing floating point ops, we haven't produced static/dynamic quantized ops at this point.

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support dynamic quantization without calibration"


Summary:
Run weight observer for dynamic quantization before inserting quant/dequant node

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23452123](https://our.internmc.facebook.com/intern/diff/D23452123)

[ghstack-poisoned]
